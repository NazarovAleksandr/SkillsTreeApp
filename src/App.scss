@import './variables.scss';

#root, body, html {
    height: 100%;
    margin: 0;
    padding: 0;
    overflow: hidden;
    background-color: black;
}

.App {
    font-family: Arial, Helvetica, sans-serif;
    height: 100%;
    display: flex;
    margin: auto;
    max-width: 1650px;

    .layout {
        display: flex;
        flex-direction: column;
        flex-grow: 1;
        margin: 3%;

        .pages {
            flex-grow: 1;
            display: flex;
            overflow: auto;
        }
    }

    .menu {
        border-bottom: 2px solid $mainBorderColor;
        margin-bottom: 10px;
        text-align: center;

        > a {
            display: inline-block;
            margin-right: 20px;
            padding: 5px;
            color: #faa318;
            text-decoration: none;
            width: 100px;

            &:last-child {
                margin-right: 0;
            }

            &.active, &:hover {
                box-shadow: 0px 2px #faa318;
            }
        }
    }
}