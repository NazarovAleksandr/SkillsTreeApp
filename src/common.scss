@import './variables.scss';

.rune-socket {
    width: 64px;
    height: 64px;
    display: inline-block;
    border: 2px solid;
    border-radius: 40px;
    background: #171717;
    vertical-align: middle;
    box-shadow: 0 2px #000;
    margin: 8px;

    &.unique {
        color: $uniqueRarity;
    }

    &.common {
        color: $commonRarity;
    }

    &.rare {
        color: $rareRarity;
    }
}

.rune {
    &-image {
        width: 64px;
        height: 64px;
    }

    &.used {
        background-color: yellow;
    }
}

.block-title {
    border-bottom: 2px solid $mainBorderColor;
    margin-bottom: 10px;
    padding: 10px;
    color: #faa318;

    .title {
        padding: 6px;
        display: inline-block;
        text-transform: uppercase;
    }
}

.action {
    color: rgb(250, 163, 24);
    border: 1px solid #545454;
    display: inline-block;
    padding: 5px;
    box-shadow: 2px 2px #000;
    text-align: center;
    cursor: default;
    position: relative;
    user-select: none;

    &.disabled {
        color: gray;
    }

    &:hover:not(.disabled) {
        border-color: rgb(250, 163, 24);
        cursor: pointer;
    }
}

@keyframes drop-splash {
	0% {
        width: 70px;
        height: 70px;
		opacity: 1;
		transform: scale3d(0.5, 0.5, 1);
	}
	100% {
		opacity: 0;
        transform: scale3d(1.1, 1.1, 1);
        width: 70px;
        height: 70px;
	}
}

@keyframes drop-splash-2 {
	0% {
        width: 70px;
        height: 70px;
		opacity: 1;
		transform: scale3d(0.5, 0.5, 1);
	}
	50%, 100% {
        width: 70px;
        height: 70px;
		opacity: 0;
		transform: scale3d(1.2, 1.2, 1);
	}
}

@keyframes rotation {
    100% {
        transform: rotate(360deg);
    }
}