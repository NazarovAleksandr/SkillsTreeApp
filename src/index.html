<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8"/>
    <meta name="viewport" content="initial-scale=0.6, shrink-to-fit=yes">
    <title>Game proof-of-concept</title>
  </head>
  <body>
    <div id="root"></div>
</html>