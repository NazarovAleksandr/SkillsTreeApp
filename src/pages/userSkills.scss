.user-skills-page {
    display: flex;
    flex-grow: 1;
    position: relative;

    

    &:before {
        content: "";
        position: absolute;
        top: 0; left: 0;
        width: 100%; height: 100%;
        background-repeat: no-repeat;
        background-position: right -300px center;
        background-image: url('../assets/images/wall.jpg');
        filter: blur(2px);
        z-index: 1;
    }
}