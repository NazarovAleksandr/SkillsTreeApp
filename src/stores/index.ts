import * as magicSchools from './magicSchool';
import * as runes from './runes';
import * as skillTrees from './skilltrees';
import * as tooltipState from './tooltipState';

export {
    magicSchools,
    runes,
    skillTrees,
    tooltipState
}