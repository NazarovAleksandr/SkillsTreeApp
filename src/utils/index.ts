import * as randomizer from './randomizer';
import * as uids from './uids';
import * as constants from './constants';
import { images } from './images';

export {
    randomizer,
    uids,
    constants,
    images,
};
