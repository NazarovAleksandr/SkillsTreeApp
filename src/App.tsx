import * as React from 'react';
import './App.scss';
import './common.scss';
import { observer } from "mobx-react";
import AboutPage from './pages/about';
import UserSkillsPage from './pages/userSkills';
import { BrowserRouter as Router, Route, NavLink } from "react-router-dom";
import { TooltipContext } from './contexts/tooltip';
import * as stores from './stores';
import { Tooltip } from './components/tooltip/tooltip';

const tooltipStateStore = new stores.tooltipState.TooltipStateStore();
let basename = process.env.BASENAME || '/'

window.addEventListener('scroll', () => {
    tooltipStateStore.hide();
}, {
    capture: true,
    passive: true
});

const App = () => {
    return (
        <Router basename={basename}>
            <TooltipContext.Provider value={tooltipStateStore}>
                <div className="App">
                    <div className="layout">
                        <div className="menu">
                            <NavLink exact to="/" activeClassName='active'>About</NavLink>
                            <NavLink to="/skills" activeClassName='active'>Skills</NavLink>
                        </div>
                        <div className="pages">
                            <Route path="/skills" component={UserSkillsPage} />
                            <Route exact path="/" component={AboutPage} />
                        </div>
                    </div>
                </div>
            </TooltipContext.Provider>
            <Tooltip 
                position={tooltipStateStore.tooltipPosition}
                content={tooltipStateStore.tooltipContent}
                node={tooltipStateStore.hostNode}
                visible={tooltipStateStore.isVisible}
                mouseOutCallback={tooltipStateStore.mouseOutCallback}
            />
        </Router>
    );
}

export default observer(App);

/*

TODO

1) Рефакторинг

*/