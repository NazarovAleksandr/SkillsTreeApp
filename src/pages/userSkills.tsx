import './userSkills.scss'
import * as React from 'react';
import { observer } from "mobx-react";
import { observable } from "mobx";
import MagicSchools from '../components/magicSchools/magicSchools';
import SkillsTree from '../components/skillsTree/skillsTree';
import RunesList from '../components/runes/runesList';
import * as Models from '../components/magicSchools/models';
import {IUserSkillsPageProps} from './models';
import { DragDropContext, DropResult } from 'react-beautiful-dnd';
import * as utils from './../utils';



@observer
export class UserSkillsPage extends React.PureComponent<IUserSkillsPageProps> {
    @observable
    public activeSchool: Models.IMagicSchool;

    public componentDidMount() {
        this.setActiveSchool(this.props.uiStateStore.getSelectedSchoolId());
    }

    private setActiveSchool = (schoolId: string) => {
        this.activeSchool = this.props.magicSchoolsStore.getSchool(schoolId);
    }

    public onRuneDropped = (result: DropResult) => {
        const { source, destination } = result;
        if (source && destination) {
            let rune = this.props.runesStore.getRune(source.droppableId);
            let node = this.props.skillTreesStore.getNode(destination.droppableId);
    
            if (rune && node) {
                node.attachedRune = rune;
                navigator.vibrate && navigator.vibrate([100]);
            }
        }
        document.body.classList.remove(utils.constants.dragInProgress);
    }

    public onDragStart() {
        document.body.classList.add(utils.constants.dragInProgress);
        navigator.vibrate && navigator.vibrate([100]);
    }

    public render() {
        let {magicSchoolsStore, uiStateStore, skillTreesStore, runesStore} = this.props;
        return (
            <div className="user-skills-page">
                <MagicSchools schoolsStore={magicSchoolsStore} uiStateStore={uiStateStore}></MagicSchools>
                <DragDropContext onDragEnd={this.onRuneDropped} onDragStart={this.onDragStart}>
                    <SkillsTree relatedSchoolId={uiStateStore.getSelectedSchoolId()} treeStore={skillTreesStore}></SkillsTree>
                    <RunesList runesStore={runesStore} usedRunes={skillTreesStore.attachedRunes}></RunesList>
                </DragDropContext>
            </div>
        );
        
    }
}