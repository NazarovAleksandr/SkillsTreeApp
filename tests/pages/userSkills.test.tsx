import {UserSkillsPage} from '../../src/pages/userSkills';
import React = require('react');
import { shallow } from 'enzyme';
import * as stores from '../../src/stores';
import * as utils from '../../src/utils';

let magicSchoolsStore: stores.magicSchools.MagicSchoolsStore;
let skillTreesStore: stores.skillTrees.SkillTreesStore;
let uiStateStore: stores.magicSchools.MagicSchoolsUIStateStore;
let runesStore: stores.runes.RunesStore;

beforeEach(() => {
    magicSchoolsStore = new stores.magicSchools.MagicSchoolsStore();
    skillTreesStore = new stores.skillTrees.SkillTreesStore();
    uiStateStore = new stores.magicSchools.MagicSchoolsUIStateStore();
    runesStore = new stores.runes.RunesStore();
    document.body.classList.remove(utils.constants.dragInProgress);
});

describe('UserSkills page', () => {
    test('Renders', () => {
        const component = shallow(<UserSkillsPage
            magicSchoolsStore={magicSchoolsStore}
            skillTreesStore={skillTreesStore}
            uiStateStore={uiStateStore}
            runesStore={runesStore}></UserSkillsPage>);

        expect(component).toMatchSnapshot();
    });
    test('Sets active school', () => {
        let school = magicSchoolsStore.generateSchool();
        magicSchoolsStore.addSchool(school);
        uiStateStore.setActiveSchool(school.id);

        const component = shallow<UserSkillsPage>(<UserSkillsPage
            magicSchoolsStore={magicSchoolsStore}
            skillTreesStore={skillTreesStore}
            uiStateStore={uiStateStore}
            runesStore={runesStore}></UserSkillsPage>);

        let activeSchool = component.instance().activeSchool;

        expect(activeSchool).toEqual(school);
    });
    test('Sets class to body while dragging', () => {
        const component = shallow<UserSkillsPage>(<UserSkillsPage
            magicSchoolsStore={magicSchoolsStore}
            skillTreesStore={skillTreesStore}
            uiStateStore={uiStateStore}
            runesStore={runesStore}></UserSkillsPage>);

        component.instance().onDragStart();
        expect(document.body.classList.contains(utils.constants.dragInProgress)).toEqual(true);
    });
    test('Removes class from body after dropping', () => {
        const component = shallow<UserSkillsPage>(<UserSkillsPage
            magicSchoolsStore={magicSchoolsStore}
            skillTreesStore={skillTreesStore}
            uiStateStore={uiStateStore}
            runesStore={runesStore}></UserSkillsPage>);

        component.instance().onRuneDropped({
            combine: undefined,
            destination: undefined,
            draggableId: undefined,
            mode: undefined,
            reason: undefined,
            source: undefined,
            type: undefined
        });
        expect(document.body.classList.contains(utils.constants.dragInProgress)).toEqual(false);
    });
    test('Attaches rune after dropping to node', () => {
        const component = shallow<UserSkillsPage>(<UserSkillsPage
            magicSchoolsStore={magicSchoolsStore}
            skillTreesStore={skillTreesStore}
            uiStateStore={uiStateStore}
            runesStore={runesStore}></UserSkillsPage>);
        
        let rune = runesStore.generateRune();
        runesStore.addRune(rune);
        let tree = {
            id: '1'
        };
        skillTreesStore.addTree('1', tree);

        component.instance().onRuneDropped({
            combine: undefined,
            destination: {droppableId: '1', index: 1},
            draggableId: undefined,
            mode: undefined,
            reason: undefined,
            source: {droppableId: rune.id, index: 1},
            type: undefined
        });
        
        expect(skillTreesStore.getNode('1').attachedRune).toEqual(rune);
    });
});