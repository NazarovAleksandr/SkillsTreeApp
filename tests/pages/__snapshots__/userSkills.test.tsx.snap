// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSkills page Renders 1`] = `
<div
  className="user-skills-page"
>
  <MagicSchools
    schoolsStore={
      MagicSchoolsStore {
        "generateSchool": [Function],
        "schools": Array [],
      }
    }
    uiStateStore={
      MagicSchoolsUIStateStore {
        "inEditionState": undefined,
        "selectedSchoolId": undefined,
      }
    }
  />
  <DragDropContext
    onDragEnd={[Function]}
    onDragStart={[Function]}
  >
    <SkillsTree
      treeStore={
        SkillTreesStore {
          "schoolIdToTree": Object {},
          "trees": Array [],
        }
      }
    />
    <RunesList
      runesStore={
        RunesStore {
          "generateImage": [Function],
          "getRandomEntity": [Function],
          "propertyTypes": Array [
            "Evade",
            "Health",
            "Mana",
            "Power",
            "Shield",
          ],
          "runeRarities": Array [
            "Common",
            "Rare",
            "Unique",
          ],
          "runeTypes": Array [
            "Attack",
            "Defence",
            "Heal",
            "Magic",
            "Utility",
          ],
          "runes": Array [],
        }
      }
      usedRunes={Set {}}
    />
  </DragDropContext>
</div>
`;
